### DISCUSSION TITLE
What's the best way to deploy a Next app with a Python server?

### DISCUSSION THREAD

**Discussion by tylerbecks** [vercel/next.js#15846] [Help] [answered]
NOTE: I've never deployed a python server before, but I have a working python script.  I want to create an API endpoint using my code for my Next app.

I'm looking for the best/easiest way to deploy a python server with my Next app. At first, I thought I could add it to `pages/api/index.py` but it seems that only supports js and ts files.

So next I thought I would try a root level `api/index.py`folder, but after running `yarn run dev` and visiting http://localhost:3000/api, I get a 404.  Has anyone done this? What is your suggestion?

I saw this issue on github, but it's unresolved:
https://github.com/vercel/next.js/issues/10770

**Comment by jamesmosier** [comment, marked as answer]
Next has no way to handle Python code, so you would have to host your Python API separately and then use something like a [rewrite](https://nextjs.org/docs/api-reference/next.config.js/rewrites) to point a route to this Python API route.

```js
module.exports = {
  async rewrites() {
    return [
      {
        source: '/api/my-api-route',
        destination: '/python-api/some-route',
      },
    ]
  },
}
```

**Reply by tylerbecks** [reply]
Thanks for the tip! Any suggestions on how/where to host my Python API?

**Reply by lagupa** [reply]
This is probably late, but for something like this, I would create my API with DjangoRestFramework and deploy it on platforms like digital ocean or Heroku, or AWS. The API can then be called from the next js app.

**Reply by bmarianome** [reply]
You cannot make post requests from next.js if the api is in a different domain, you would have to deactivate csrf protection

**Reply by eric-burel** [reply]
Can you elaborate on this @marianobenitezz ? I can't see why, except that you need to setup CORS correctly on the API to accept the client request
Also this answer is very outdated, see my answer below, Python API is doable in Next

**Reply by bmarianome** [reply]
I tried to make an user authentication with nextjs and django, but to send a post request to django, you need to include the csrf token in the headers. How can the nextjs app have the django-provided csrf token in a different domain?
I saw this answer here that says it is not possible: https://stackoverflow.com/a/22002765/18289248

**Reply by GabenGar** [reply]
> How can the nextjs app have the django-provided csrf token in a different domain?

By querying django endpoint for CSRF token before sending a response with a form.

**Comment by eric-burel** [comment]
Hi guys, I am giving a shot at allowing Python API routes within a Next.js app. Because that would be the best thing to happen for data visualization since the invention of the bar chart.
Link here: https://github.com/lbke/snext

This doesn't really work yet, but I've documented the various blocker that would need to be solved to achieve the same experience with Python API routes than with JS/TS routes: https://github.com/lbke/snext#main-blockers

**Reply by eric-burel** [reply]
Ok actually after some digging, it seems to work ok with the following workflow:
- Use `./api` folder **at the root** of your app, instead of nesting it in `pages/`
- Use `vercel dev` instead of `next dev`

Boom, works like a charm. You can have Python API routes alongside Node.js API routes (or whatever language).

It was described in this ticket as well: https://github.com/vercel/vercel/discussions/4023

**Reply by nikhilmeria** [reply]
> 
> 
> Ok actually after some digging, it seems to work ok with the following workflow:
> 
>     * Use `./api` folder **at the root** of your app, instead of nesting it in `pages/`
> 
>     * Use `vercel dev` instead of `next dev`
> 
> 
> Boom, works like a charm. You can have Python API routes alongside Node.js API routes (or whatever language).
> 
> It was described in this ticket as well: [vercel/vercel#4023](https://github.com/vercel/vercel/discussions/4023)

can I hv ur email id? need to consultant you on this topic. thank you
